pkg/rule/postgres.go

```diff
@@ -0,0 +1,498 @@
+package rule
+
+import (
+	"database/sql"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/jmoiron/sqlx"
+
+	"github.com/lifesum/configsum/pkg/errors"
+	"github.com/lifesum/configsum/pkg/pg"
+)
+
+const (
+	pgRuleCreateSchema = `CREATE SCHEMA IF NOT EXISTS rule`
+	pgRuleCreateTable  = `
+		CREATE TABLE IF NOT EXISTS rule.rules(
+			id TEXT NOT NULL PRIMARY KEY,
+			active BOOLEAN NOT NULL,
+			activated_at TIMESTAMP WITHOUT TIME ZONE DEFAULT (now() AT TIME ZONE 'utc'),
+			buckets JSONB NOT NULL,
+			config_id TEXT NOT NULL,	
+			created_at TIMESTAMP WITHOUT TIME ZONE DEFAULT (now() AT TIME ZONE 'utc'),
+			criteria JSONB NOT NULL,
+			description TEXT NOT NULL,
+			end_time TIMESTAMP WITHOUT TIME ZONE DEFAULT (now() AT TIME ZONE 'utc'),
+			kind INT8 NOT NULL,
+			name TEXT NOT NULL,
+			start_time TIMESTAMP WITHOUT TIME ZONE DEFAULT (now() AT TIME ZONE 'utc'),
+			updated_at TIMESTAMP WITHOUT TIME ZONE DEFAULT (now() AT TIME ZONE 'utc')
+		)`
+	pgRuleDropTable = `DROP TABLE IF EXISTS rule.rules CASCADE`
+
+	pgRuleInsert = `
+		INSERT INTO
+		rule.rules(
+			id, 
+			active, 
+			activated_at, 
+			buckets, 
+			config_id, 
+			created_at, 
+			criteria, 
+			description,
+			end_time,
+			kind,
+			name,
+			start_time, 
+			updated_at) 
+			VALUES(
+				:id,
+				:active,
+				:activatedAt,
+				:buckets,
+				:configId,
+				:createdAt,
+				:criteria,
+				:description,
+				:endTime,
+				:kind,
+				:name,
+				:startTime,
+				:updatedAt
+		)`
+
+	pgRuleGet = `
+		SELECT
+			id, 
+			active, 
+			activated_at, 
+			buckets, 
+			config_id, 
+			created_at, 
+			criteria, 
+			description, 
+			end_time, 
+			kind, 
+			name, 
+			start_time, 
+			updated_at
+		FROM
+			rule.rules
+		WHERE
+			config_id = :configId	
+			AND name = :name
+		ORDER BY
+			created_at DESC
+		LIMIT
+			1`
+
+	pgRuleUpdate = `
+		UPDATE rule.rules
+		SET
+			active = :active,
+			activated_at = :activatedAt,
+			buckets = :buckets,
+			criteria = :criteria,
+			description = :description,
+			end_time = :endTime,
+			kind = :kind,
+			name = :name,
+			start_time = :startTime,
+			updated_at = :updatedAt
+		WHERE
+			config_id = :configId
+			AND name = :name
+	`
+
+	pgRuleListAll = `
+		SELECT 	
+			*
```

Comment:
Selecting all fields is better to be avoided as it doesn't control what data is fetched, which can lead to runtime errors, in cases where the shape of the data changed but the code path deserialisaling it still operates under different assumptions.

Beyond that it might have performance profiling implications.